src/Sylius/Bundle/CoreBundle/Provider/CatalogPromotionProductsProvider.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Provider;
+
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Core\Provider\CatalogPromotionProductsProviderInterface;
+use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
+
+final class CatalogPromotionProductsProvider implements CatalogPromotionProductsProviderInterface
```

Comment:
imo it should be named `CatalogPromotionVariantsProvider`, CPs are applied on the variant, and this service provide variants :)

Reply:
But this service provides products 😄 Of course, the discount is applied on variants, but it's done in `CatalogPromotionApplicator`. Here we get products that contains variants specified in the rule (as the rule name says so)

Reply:
But why are you getting products that contain variants? We should just apply discounts on variants specified in catalog promotion rule and then, this service is not needed

Reply:
After some consideration - yes, I wanted to be **too** consistent with the regular promotions concept 🖖 Gonna fix it in a minute